contributors/design-proposals/storage/dynamic_volume_limit.md

```diff
@@ -0,0 +1,157 @@
+# Dynamic attached volume limits
+
+## Goals
+
+Currently the number of volumes attachable to a node is either hard-coded or only configurable via an environment variable. Also
+existing limits only apply to well known volume types like EBS, GCE and is not available to all volume plugins.
+
+This proposal enables any volume plugin to specify those limits and also allows same volume type to have different volume
+limits depending on type of node.
+
+## Implementation Design
+
+### Prerequisite
+
+* 1.11 This feature will be protected by an alpha feature gate, so as API and CLI changes needed for it. We are planning to call
+  the feature `AttachVolumeLimit`.
+* 1.12 This feature will be behind a beta feature gate and enabled by default.
+
+### API Changes
+
+There is no API change needed for this feature. However existing `node.Status.Capacity` and `node.Status.Allocatable` will
+be extended to cover volume limits available on the node too.
+
+The key name that will store volume will be start with prefix `attachable-volumes-`. The volume limit key will respect
+format restrictions applied to Kubernetes Resource names. Volume limit key for existing plugins might look like:
+
+
+* `attachable-volumes-aws-ebs`
```

Comment:
I really dislike this sort of ad hoc syntax embedded in a string, and aside from that, I don't buy that this should be magical like this.  Why can't we do something less stringy?

Just thinking out loud...

e.g. Some volume have per-node attachment limits.  For those volumes, you install an admission controller that examines the volumes in play and decorates the pod as requiring certain resources, and then the scheduler simply schedules?

I feel like we have explored this, but if so, I don't recall why it was rejected.

Reply:
There are two part of it right?

1. The plugin must be able to specify a maximum attachment limit on per node basis. We chose to use node's allocatable/capacity field for that. And when we want to express this limit in node's allocatable regardless of how we do counting via admission controller or something else - the limit must have a key and a value. The key MUST be a string that adheres to resourceName naming conventions(allocatable is of type `map[v1.ResourceName]resource.Quantity`).  So isn't the `attachable-volumes-aws-ebs` prefix chosen for expressing attach limits in node's allocatable/capacity orthogonal to how we do counting?

2. We could still use a admission controller to do actual counting but I think it was agreed that, scheduler is better(or may be easier) place for counting actual volumes in use. Because:
a. It can consider all pods on the node. Not just the one being scheduled.
b. I think the way volume limits interacts with late binding and even distribution of pods with volume workloads(https://github.com/kubernetes/kubernetes/pull/60525) makes it better place to do in scheduler. 

And obviously a admission plugin may be disabled or not enabled(but same goes for a scheduler predicate I guess). I am not particularly opposed to using admission controller for counting but I think we did not see benefits we thought we will see.

Reply:
I am going to reformat for comments:

`We chose to use node's allocatable/capacity field for that [to specify a maximum attachment limit]` ... `The key MUST be a string that adheres to resourceName naming conventions` ... `So isn't the attachable-volumes-aws-ebs prefix chosen for expressing attach limits in node's allocatable/capacity orthogonal to how we do counting`

Yes it is orthogonal.  Sorry for putting them in one message.

My first grump here is that we have invented non-obvious syntax without any specification for it.  I detest string-embedded syntaxes (stringly type APIs) and even if we DO go that way, we can't tolerate under-specified ones.  Where do these magical names come from?  How do they not conflict?  Jordan had some comments on that, too.  At the very very least this needs a strong spec, and perhaps we actually need something different.

Regarding scheduler, what I really don't want to do is build up a body of special -cases that we dump on the scheduler, when the problem is actually a pretty general problem.  How do we opaquely allocate and schedule resources WITHOUT changing the scheduler?   We may need to make other changes to the system to accomodate that, but I think it should be a preferred choice.  @bsalamat am I over-blowing this?

Late-binding is from a PVC to an PV via a class.  Presumably a single class can't use multiple drivers, so it's still feasible to reify the PVC -> class -> driver somewhere?  The fact that classes could, in theory change, is a problem that maybe we should think harder about and maybe we need a different kind of plugin (e.g. maybe we need explicit scheduler pre-processor plugins). "Let's just hack the scheduler' seems like a very slippery slope to me.

And, BTW, I *want* this capability (and have for years), so please don't think I am poo-pooing the idea :)

Reply:
I second what @thockin said about dumping a lot of special case logic in the scheduler. That said, not everything can be checked at admission time. For example, dynamic volume binding happens after scheduler finds a node to run a pod. I guess there is no choice other than checking the number of attached volumes to the node at that point.

Reply:
> Where do these magical names come from? How do they not conflict? Jordan had some comments on that, too. At the very very least this needs a strong spec, and perhaps we actually need something different.

In current iteration any resource name(in node's allocatable) that uses `attachable-volumes-` prefix refers to attachable limits. We have added some tests to ensure that containers themselves aren't able to directly request this resource.  It is however possible that someone can set a new `attachable-volumes-xxx` value that does not apply to volume limits(via `patch` ).

I am not entirely sure about some external component inadvertently overriding any of existing limits though. Because - when kubelet starts, it sets those limits and it periodically syncs the value from same source. So if an external component does override those values it will be wiped out. 

I still see the need of speccing them better and perhaps there is a better way. I need to think and will bring this up on sig-storage.

Reply:
Previous discussion [here](https://github.com/kubernetes/community/pull/2051#discussion_r183839230)

Reply:
Just had a discussion with @jsafrane as well about this and we discussed two problematic things about moving the count outside the scheduler:

1. For local volumes with delayed binding, storageClass could be dummy or unknown and hence counting unbound PVCs is going to be tricky outside the scheduler.
2. Currently the scheduler counts unique volumes. If we specify volumes as a pod level countable resource and scheduler counts volumes the way it counts memory or other resources, it can't accurately count unique volumes.

Reply:
> It is however possible that someone can set a new attachable-volumes-xxx value that does not apply to volume limits(via patch ).

What is the specification for xxx here?  Is it the CSI driver name?  Some other source?  Does it get some auto-mutation?  Are characters changed?